osdep/io.c

```diff
@@ -894,4 +883,95 @@ void freelocale(locale_t locobj)
 {
 }
 
+#define MP_PIPE_BUF_SIZE 65536
+
+int mp_make_cloexec_pipe(int pipes[2])
+{
+    if (_pipe(pipes, MP_PIPE_BUF_SIZE, _O_BINARY | _O_NOINHERIT) != 0) {
+        pipes[0] = pipes[1] = -1;
+        return -1;
+    }
+    return 0;
+}
+
+int mp_make_wakeup_pipe(int pipes[2])
+{
+    static atomic_ulong pipe_id = 0;
+
+    HANDLE readh = INVALID_HANDLE_VALUE;
+    HANDLE writeh = INVALID_HANDLE_VALUE;
+    pipes[0] = pipes[1] = -1;
+
+    char pipe_name[64];
+    unsigned long pid = mp_getpid();
+    sprintf(pipe_name, "\\\\?\\pipe\\mpv\\%lu-%lu", pid, atomic_fetch_add_explicit(&pipe_id, 1, memory_order_relaxed));
+
+    writeh = CreateNamedPipeA(
+        pipe_name,
+        PIPE_ACCESS_OUTBOUND | FILE_FLAG_FIRST_PIPE_INSTANCE | FILE_FLAG_OVERLAPPED,
+        0,
+        1,
+        MP_PIPE_BUF_SIZE,
+        MP_PIPE_BUF_SIZE,
+        0,
+        NULL
+    );
+    if (writeh == INVALID_HANDLE_VALUE) {
+        set_errno_from_lasterror();
+        goto error;
+    }
+
+    readh = CreateFileA(
+        pipe_name,
+        GENERIC_READ,
+        0,
+        NULL,
+        OPEN_EXISTING,
+        FILE_FLAG_OVERLAPPED,
+        NULL
+    );
+    if (readh == INVALID_HANDLE_VALUE) {
+        set_errno_from_lasterror();
+        goto error;
+    }
+
+    if (!ConnectNamedPipe(writeh, NULL) && GetLastError() != ERROR_PIPE_CONNECTED) {
+        set_errno_from_lasterror();
+        goto error;
+    }
+
+    pipes[0] = _open_osfhandle((intptr_t)readh, 0);
+    if (pipes[0] == -1)
+        goto error;
+    pipes[1] = _open_osfhandle((intptr_t)writeh, 0);
+    if (pipes[1] == -1)
+        goto error;
+    return 0;
+
+error:
+    if (pipes[0] != -1)
+        _close(pipes[0]);
+    else if (readh != INVALID_HANDLE_VALUE)
+        CloseHandle(readh);
+    if (pipes[1] != -1)
+        _close(pipes[1]);
+    else if (writeh != INVALID_HANDLE_VALUE)
+        CloseHandle(writeh);
+    pipes[0] = pipes[1] = -1;
+    return -1;
+}
+
+void mp_flush_wakeup_pipe(int pipe_end)
+{
+    char buf[100];
+    OVERLAPPED operation = {};
+    HANDLE handle = (HANDLE)_get_osfhandle(pipe_end);
```

Comment:
Why not read `pipe_end` directly? Do we need to use winapi here?

Reply:
Unfortunatly, `_read` will block.